system-addon/lib/MessageCenterRouter.jsm

```diff
@@ -0,0 +1,204 @@
+  }
+
+  async onMessage({data: action, target}) {
+    switch (action.type) {
+      case "CONNECT_UI_REQUEST":
+      case "GET_NEXT_MESSAGE":
+        await this.sendNextMessage(target);
+        break;
+      case "BLOCK_MESSAGE_BY_ID":
+        await this.setState(state => {
+          const newState = Object.assign({}, state.blockList);
+          newState[action.data.id] = true;
+          return {blockList: newState};
+        });
+        this.clearMessage(target, action.data.id);
+        break;
+      case "UNBLOCK_MESSAGE_BY_ID":
+        await this.setState(state => {
+          const newState = Object.assign({}, state.blockList);
+          delete newState[action.data.id];
+          return {blockList: newState};
+        });
+        break;
+      case "ADMIN_CONNECT_STATE":
+        target.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "ADMIN_SET_STATE", data: this.state});
+        break;
+    }
+  }
+
+  /**
+   * init - Initializes the MessageRouter.
+   * It is ready when it has been connected to a RemotePageManager instance.
+   *
+   * @param {RemotePageManager} channel a RemotePageManager instance
+   * @memberof _MessageCenterRouter
+   */
+  init(channel) {
+    this.messageChannel = channel;
+    this.messageChannel.addMessageListener(INCOMING_MESSAGE_NAME, this.onMessage);
+    this.initialized = true;
+  }
+
+  uninit() {
+    this.messageChannel.sendAsyncMessage(OUTGOING_MESSAGE_NAME, {type: "CLEAR_MESSAGE"});
+    this.messageChannel.removeMessageListener(INCOMING_MESSAGE_NAME, this.onMessage);
+    this.messageChannel = null;
+    this.initialized = false;
+  }
+}
+this._MessageCenterRouter = _MessageCenterRouter;
+
+/**
+ * MessageCenterRouter - singleton instance of _MessageCenterRouter that controls all messages
+ * in the new tab page.
+ */
+const MessageCenterRouter = new _MessageCenterRouter({messages: FAKE_MESSAGES});
+
+/**
+ * @class MessageCenterFeed - Connects MessageCenter singleton (see above) to Activity Stream's
+ * store so that it can use the RemotePageManager.
+ */
+class MessageCenterFeed {
+  constructor(options = {router: MessageCenterRouter}) {
+    this.router = options.router;
+  }
+
+  /**
+   * enableOrDisableBasedOnPref - Check the experiment pref
+   * (messageCenterExperimentEnabled) and enable or disable MessageCenter based on
+   * its value.
+   */
+  enableOrDisableBasedOnPref() {
+    const isExperimentEnabled = this.store.getState().Prefs.values.messageCenterExperimentEnabled;
+    if (!this.router.initialized && isExperimentEnabled) {
+      this.router.init(this.store._messageChannel.channel);
+    } else if (!isExperimentEnabled && this.router.initialized) {
+      this.router.uninit();
+    }
+  }
+
+  onAction(action) {
+    switch (action.type) {
+      case at.INIT:
+      case at.PREF_CHANGED:
+        this.enableOrDisableBasedOnPref();
+        break;
+      case at.UNINIT:
+        this.router.uninit();
```

Comment:
If you toggle off the experiment and shut down the browser `uninit` will be called twice.